24.md

```diff
@@ -0,0 +1,395 @@
+
+Direct Message Payload: 
+```js
+{
+  "id": "<The usual hash>",
+  "pubkey": "<author key in hex>",
+  "content": "This is a DM",
+  "kind": 14,
+  "created_at": 1686840217,
+  "tags": [ <one or more receivers, mentions, citations, additional fields like content-sensitivity, etc> ],
+}
+```
+
+## 2. The Sealed Gossip Event Kind
+
+A Sealed Gossip is a `kind:13` event that wraps an unsigned event (the Gossip) with the sender's regular key. The Sealed Gossip is **always** encrypted to a receiver's pubkey but there is no `p` tag pointing to the receiver. There is no way to know who the Gossip is for without the receiver's or the sender's private key. The only public information in this event is who is signing it. 
+
+```js
+{
+  "id": "<The usual hash>",
+  "pubkey": "<Real Author's PubKey>",
+  "content": "<XCHACHA-Encrypted Gossip Kind>",
+  "kind": 13,
+  "created_at": 1686840217,
+  "tags": [],
+  "sig": "<Real Author's PubKey Signature>"
+}
+```
+
+Tags MUST must always be empty in a `kind:13`, 
+
+The inner event MUST always be unsigned 
+
+The encryption algorithm MUST use a versioned payload with a XChaCha cipher. 
+
+## 3. Gift Wrap Event Kind
+
+A GiftWrap event is a `kind:1059` event that wraps any other event with a single use (random) key and is `p`-Tag-addressed to the receiver.
+
+The goal is to hide the sender's information, the metadata, and the content of the original event from the public. The only public information is the receiver's public key. Clients pull all GiftWraps citing their user, decrypt the gift-wrapped event with the user's private key, and use the inner event as usual. The inner event is signed by the known author. 
+
+Gift wraps alone are an effective tool to hide information from the public. However, the inner event can be easily re-broadcasted, breaking the privacy of the sender and leaking its metadata (e.g. author, event kind, dates, tags, etc). The GiftWrap MUST wrap a Sealed Gossip Event to make sure the receiver cannot break the privacy guarantees when re-broadcasting it. 
+
+```js
+{
+  "id": "<The usual hash>",
+  "pubkey": "<Random, one-time-use PubKey>",
+  "content": "<XCHACHA-Encrypted Kind 13>",
+  "kind": 1059,
+  "created_at": 1686840217,
+  "tags": [
+    [ "p", "<Receiver>" ]
+  ],
+  "sig": "<Random, one-time-use PubKey Signature>"
+}
+```
+
+A `p` tag for the recipient MUST be added to the `tags` field of a GiftWrap. This is how clients will download incoming "things". Other tags should not be added to this event. 
+
+## 4. Encrypting with XChaCha (former NIP-44)
+
+To wrap an event for Sealed Gossip or GiftWrap, first serialize it to JSON. Then acquire a shared secret between the signing private key and the recipient's public key, a 24-byte nonce, and encrypt it with XChaCha (XChaCha is different than ChaCha). Convert both the nonce and the resulting byte array to Base64 and output an object with the following fields: 
```

Comment:
maybe csv?

`1,base64(ciphertext),base64(nonce)` vs 
`{ "ciphertext": "<Base64-encoded ByteArray>", "nonce": "<Base64-encoded Nonce>", "v": <Version Code>  }`

Reply:
We already have a JSON parser in so many other places. Adding one more wouldn't change much. I don't think we have a CSV parser yet, so I would side with the JSON payload for now.

Reply:
Which other things are using json?

Also, there's no need to conform to CSV spec -- just using basic `a,b,c` seems to work.

My concern here is not JSON itself, which is great, but the fact that it's nested JSON. E.g. JSON.parse(JSON.parse(response).content), which can be its own source of errors.

Reply:
Also, using JSON would increase every message by 30 bytes. Base64 is only used because it saves a lot of traffic.

Readability is cool, but the fields are fixed for every algo version:

```
1,base64(nonce),base64(ciphertext) # always same
2,base64(nonce),base64(kdf_key),base64(ciphertext) # example algorithm 2, always same
```

Reply:
> Which other things are using nested json?

Reposts, Generic Reposts, and Community Post Approvals have the reposted event as a JSON string in the content. UserMetadata in Kind0 and the Relay list on Kind 3 is a JSON string. NIP28 Channel Metadata is also a JSON string. Private zaps use a JSON string to hide the private part. All NIP51 + Mute, Pin, Block lists have a private part which is a JSON string of the private tag list. Besides that, downloading Nostr events via WebSockets means parsing 1000s of JSONs a second. 

> JSON.parse(JSON.parse(response).content)

This is extremely common, though. 

> Using JSON would increase every message by 30 bytes. 

Not a big deal. Nostr encodes keys, hashes, and signatures in Hex strings, which is already wasteful. This protocol is not good for byte crunching. We should settle for what's the easiest encode/decode to facilitate adoption.

Reply:
if it is already present in other events as you've mentioned, then I have no objections to using json